From bf3bba7c180db293678191f32bcb258ac3c4607d Mon Sep 17 00:00:00 2001
From: Chris Wedgwood <cw@f00f.org>
Date: Thu, 1 Oct 2026 14:20:19 -0700
Subject: [PATCH] virt-api: accept large identity header sets with Go 1.27

Go 1.27 rejects requests with more than 500 header values by default
(http.Server.MaxHeaderValueCount, golang/go#79936). kube-apiserver
forwards each of a user's groups to aggregated API servers as a
separate X-Remote-Group header, so users with about 500 groups receive
HTTP 431 from virt-api when it is built with Go 1.27.

Set MaxHeaderValueCount to 8192 on the virt-api server for Go 1.27+
builds: 1 MiB at 128 bytes per value. Earlier toolchains use a no-op.

Signed-off-by: Chris Wedgwood <cw@f00f.org>
---
 pkg/virt-api/BUILD.bazel                |  7 ++-
 pkg/virt-api/api.go                     |  1 +
 pkg/virt-api/header_limit_go127.go      | 29 +++++++++++++
 pkg/virt-api/header_limit_go127_test.go | 57 +++++++++++++++++++++++++
 pkg/virt-api/header_limit_pre_go127.go  | 26 +++++++++++
 5 files changed, 119 insertions(+), 1 deletion(-)
 create mode 100644 pkg/virt-api/header_limit_go127.go
 create mode 100644 pkg/virt-api/header_limit_go127_test.go
 create mode 100644 pkg/virt-api/header_limit_pre_go127.go

diff --git a/pkg/virt-api/BUILD.bazel b/pkg/virt-api/BUILD.bazel
index 738197de27ae..ea1035f286c1 100644
--- a/pkg/virt-api/BUILD.bazel
+++ b/pkg/virt-api/BUILD.bazel
@@ -2,7 +2,11 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "go_default_library",
-    srcs = ["api.go"],
+    srcs = [
+        "api.go",
+        "header_limit_go127.go",
+        "header_limit_pre_go127.go",
+    ],
     importpath = "kubevirt.io/kubevirt/pkg/virt-api",
     visibility = ["//visibility:public"],
     deps = [
@@ -54,6 +58,7 @@ go_test(
     name = "go_default_test",
     srcs = [
         "api_test.go",
+        "header_limit_go127_test.go",
         "virt-api_suite_test.go",
     ],
     embed = [":go_default_library"],
diff --git a/pkg/virt-api/api.go b/pkg/virt-api/api.go
index 9518f4b7af7a..7dbe78730259 100644
--- a/pkg/virt-api/api.go
+++ b/pkg/virt-api/api.go
@@ -1132,6 +1132,7 @@ func (app *virtAPIApp) startTLS(informerFactory controller.KubeInformerFactory)
 		// See CVE-2023-44487
 		TLSNextProto: map[string]func(*http.Server, *tls.Conn, http.Handler){},
 	}
+	configureIdentityHeaderLimit(server)
 
 	// start TLS server
 	go func() {
diff --git a/pkg/virt-api/header_limit_go127.go b/pkg/virt-api/header_limit_go127.go
new file mode 100644
index 000000000000..cae0f80260a1
--- /dev/null
+++ b/pkg/virt-api/header_limit_go127.go
@@ -0,0 +1,29 @@
+//go:build go1.27
+
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors.
+ *
+ */
+
+package virt_api
+
+import "net/http"
+
+func configureIdentityHeaderLimit(server *http.Server) {
+	// 1 MiB at 128 bytes per header value; allow thousands of X-Remote-Group values.
+	server.MaxHeaderValueCount = 8192
+}
diff --git a/pkg/virt-api/header_limit_go127_test.go b/pkg/virt-api/header_limit_go127_test.go
new file mode 100644
index 000000000000..92492b88911b
--- /dev/null
+++ b/pkg/virt-api/header_limit_go127_test.go
@@ -0,0 +1,57 @@
+//go:build go1.27
+
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors.
+ *
+ */
+
+package virt_api
+
+import (
+	"net/http"
+	"net/http/httptest"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("identity header limit", func() {
+	DescribeTable("accepts forwarded group headers according to the server configuration",
+		func(configure bool, wantCode int) {
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+				w.WriteHeader(http.StatusOK)
+			}))
+			if configure {
+				configureIdentityHeaderLimit(server.Config)
+			}
+			server.Start()
+			defer server.Close()
+
+			request, err := http.NewRequest(http.MethodGet, server.URL, nil)
+			Expect(err).NotTo(HaveOccurred())
+			for range 1000 {
+				request.Header.Add("X-Remote-Group", "group")
+			}
+			response, err := server.Client().Do(request)
+			Expect(err).NotTo(HaveOccurred())
+			defer response.Body.Close()
+			Expect(response.StatusCode).To(Equal(wantCode))
+		},
+		Entry("rejects with the Go default", false, http.StatusRequestHeaderFieldsTooLarge),
+		Entry("accepts with the configured limit", true, http.StatusOK),
+	)
+})
diff --git a/pkg/virt-api/header_limit_pre_go127.go b/pkg/virt-api/header_limit_pre_go127.go
new file mode 100644
index 000000000000..334e5117b406
--- /dev/null
+++ b/pkg/virt-api/header_limit_pre_go127.go
@@ -0,0 +1,26 @@
+//go:build !go1.27
+
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors.
+ *
+ */
+
+package virt_api
+
+import "net/http"
+
+func configureIdentityHeaderLimit(_ *http.Server) {}
